### PR TITLE
[S3 API]Support list objectV2 S3 API

### PR DESCRIPTION
**Pull request by SophieGuo410**
This pr is to support listObjectV2 request.

### CODE REVIEW

File: ambry-api/src/main/java/com/github/ambry/frontend/NamedBlobPath.java [hunk 1]
```diff
@@ -90,7 +91,7 @@ public static NamedBlobPath parseS3(String path, Map<String, Object> args) throw
     String[] splitPath = path.split("/", 4);
     String blobNamePrefix = RestUtils.getHeader(args, PREFIX_PARAM, false);
     boolean isGetObjectLockRequest = args.containsKey(OBJECT_LOCK_PARAM);
-    boolean isListRequest = blobNamePrefix != null;
+    boolean isListRequest = blobNamePrefix != null || "2".equals(RestUtils.getHeader(args, LIST_TYPE, false));
```
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-api/src/main/java/com/github/ambry/frontend/NamedBlobPath.java:97]: Sure, updated.

File: ambry-api/src/main/java/com/github/ambry/frontend/s3/S3MessagePayload.java [hunk 2]
```diff
@@ -222,22 +222,27 @@ public static class Contents {
     private String key;
     @JacksonXmlProperty(localName = "LastModified")
     private String lastModified;
+    @JacksonXmlProperty(localName = "Size")
```
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-api/src/main/java/com/github/ambry/frontend/s3/S3MessagePayload.java:225]: That would need to handle content separately for listObject and listObjectV2, which is not used for the current case. For now, listObject already been supported but does not include all field. But I can double check with the customer to see what they want?

File: ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java [hunk 3]
```diff
@@ -87,18 +87,18 @@ class MySqlNamedBlobDb implements NamedBlobDb {
   private static final String CURRENT_TIME = "UTC_TIMESTAMP(6)";
   private static final String STATE_MATCH = String.format("%s = %s", BLOB_STATE, NamedBlobState.READY.ordinal());
   private static final String PK_MATCH = String.format("(%s, %s, %s) = (?, ?, ?)", ACCOUNT_ID, CONTAINER_ID, BLOB_NAME);
-  private static final String PK_MATCH_VERSION = String.format("(%s, %s, %s, %s) = (?, ?, ?, ?)", ACCOUNT_ID,
-      CONTAINER_ID, BLOB_NAME, VERSION);
+  private static final String PK_MATCH_VERSION =
```
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java:90]: Sure, will try to focus I the changes next time.

File: ambry-api/src/main/java/com/github/ambry/frontend/s3/S3MessagePayload.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (JingQianCloud) [reply ambry-api/src/main/java/com/github/ambry/frontend/s3/S3MessagePayload.java:225]: The ListObject has all the fields as well.
https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjects.html
We don't need separate V1 and V2. Both are supposed to have these fields.

Just we started simple to make the client happy. But in the future I think we still need all these fields.

File: ambry-api/src/main/java/com/github/ambry/frontend/NamedBlobPath.java [hunk 4]
```diff
@@ -90,7 +91,10 @@ public static NamedBlobPath parseS3(String path, Map<String, Object> args) throw
     String[] splitPath = path.split("/", 4);
     String blobNamePrefix = RestUtils.getHeader(args, PREFIX_PARAM, false);
     boolean isGetObjectLockRequest = args.containsKey(OBJECT_LOCK_PARAM);
-    boolean isListRequest = blobNamePrefix != null;
+    //There are two cases for S3 listing
+    //1.has prefix (Ex:GET /?prefix=prefixName&delimiter=&encoding-type=url)
+    //2.no prefix but listObjectV2 (Ex:GET /?list-type=2&prefix=&delimiter=&encoding-type=url)
+    boolean isListRequest = blobNamePrefix != null || "2".equals(RestUtils.getHeader(args, LIST_TYPE, false));
```
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/frontend/NamedBlobPath.java:97]: in line 107, we are still using "page" query parameter. This is a query parameter from ambry named blob, not s3. In s3 listObject, it should be marker. And in ListObjectV2, it should be ContinuationToken.

Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/frontend/NamedBlobPath.java:97]: Please use a constant to replace "2". Something like "LIST_TYPE_VERSION_2".

File: ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3ListHandler.java [hunk 5]
```diff
@@ -99,14 +100,16 @@ private ReadableStreamChannel serializeAsXml(RestRequest restRequest, Page<Named
     String delimiter = getHeader(restRequest.getArgs(), DELIMITER_PARAM_NAME, false);
     String encodingType = getHeader(restRequest.getArgs(), ENCODING_TYPE_PARAM_NAME, false);
     String maxKeys = getHeader(restRequest.getArgs(), MAXKEYS_PARAM_NAME, false);
-    int maxKeysValue = maxKeys == null ? Integer.MAX_VALUE : Integer.parseInt(maxKeys);
+    //By default S3 list returns up to 1000 key names.
+    int maxKeysValue = maxKeys == null ? DEFAULT_MAX_KEY_VALUE : Integer.parseInt(maxKeys);
```
Comment (justinlin-linkedin) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3ListHandler.java:104]: Do we wanna change the list interface from NamedBlobDB to allow passing a maxKeyCount? In this way, we can control how many results to query in database.

File: ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java [hunk 6]
```diff
@@ -597,11 +617,17 @@ private NamedBlobRecord run_get_v2(String accountName, String containerName, Str
   private Page<NamedBlobRecord> run_list_v2(String accountName, String containerName, String blobNamePrefix,
       String pageToken, short accountId, short containerId, Connection connection) throws Exception {
     String query = "";
-    try (PreparedStatement statement = connection.prepareStatement(LIST_QUERY_V2)) {
+    try (PreparedStatement statement = blobNamePrefix == null ? connection.prepareStatement(LIST_ALL_QUERY_V2)
```
Comment (justinlin-linkedin) [inline ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java:620]: Let's do this operation on strings, not PreparedStatement. Something like.
```
String queryStatement = blobNamePrefix == null? LIST_ALL_QUERY_V2: LIST_QUERY_V2;
try (PreparedStatement state = connection.prepareStatement(queryStatement)) {
}
```

File: ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java [hunk 7]
```diff
@@ -124,6 +124,20 @@ class MySqlNamedBlobDb implements NamedBlobDb {
           + "WHERE t1.blob_name LIKE ? AND t1.blob_name >= ? ORDER BY t1.blob_name ASC LIMIT ?",STATE_MATCH, CURRENT_TIME);
   // @formatter:on
 
+  /**
+   * Similar like LIST_QUERY_V2, but this query is used when user don't provide the prefix and want to list all records.
+   */
+  // @formatter:off
+  private static final String LIST_ALL_QUERY_V2 = String.format(
+      "" + "SELECT t1.blob_name, t1.blob_id, t1.version, t1.deleted_ts " + "FROM named_blobs_v2 t1 " + "INNER JOIN "
```
Comment (justinlin-linkedin) [inline ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java:132]: the "+" character's indentation is bit off.

File: ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java [hunk 8]
```diff
@@ -124,6 +124,20 @@ class MySqlNamedBlobDb implements NamedBlobDb {
...
+          + "(SELECT account_id, container_id, blob_name, max(version) as version " + "FROM named_blobs_v2 "
+          + "WHERE (account_id, container_id) = (?, ?) AND %1$s " + "  AND (deleted_ts IS NULL OR deleted_ts>%2$S) "
+          + "        GROUP BY account_id, container_id, blob_name) t2 "
+          + "ON (t1.account_id,t1.container_id,t1.blob_name,t1.version) = (t2.account_id,t2.container_id,t2.blob_name,t2.version) "
+          + "WHERE CASE WHEN ? IS NOT NULL THEN t1.blob_name >= ? ELSE 1 END ORDER BY t1.blob_name ASC LIMIT ?",
```
Comment (justinlin-linkedin) [inline ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java:137]: Let's break this where case when then statement into multiple lines for readibility.

File: ambry-api/src/main/java/com/github/ambry/frontend/NamedBlobPath.java (hunk 4 above)
Comment (SophieGuo410) [reply ambry-api/src/main/java/com/github/ambry/frontend/NamedBlobPath.java]: Done

Comment (SophieGuo410) [reply ambry-api/src/main/java/com/github/ambry/frontend/NamedBlobPath.java]: Done

File: ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3ListHandler.java (hunk 5 above)
Comment (SophieGuo410) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3ListHandler.java:104]: I can create a follow up pr to support this.

File: ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java (hunk 6 above)
Comment (SophieGuo410) [reply ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java]: Sure, updated.

File: ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java (hunk 8 above)
Comment (SophieGuo410) [reply ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java]: Updated.

**Review by JingQianCloud — APPROVED** [review]

**Review by justinlin-linkedin — APPROVED** [review]

**Pull request merged**